### DISCUSSION TITLE
Remove or change fetchAllMembers?

### DISCUSSION THREAD

**Discussion by almostSouji** [discordjs/discord.js#5246] [Ideas]
With the introduction of gateway intents, many more users are stumbling over the `fetchAllMembers` Client option. However, in its current state, this causes major delays in ready states or even causes timeout errors.

We already discourage the use of this option in the documentation and active support in the discord. Maybe it's time to completely remove it and leave fetching members from guilds to fill caches (if necessary) to the user.

**Comment by MattIPv4** [comment]
👍 +1 for the removal of this from the client opts. It seems to cause a huge number of issues for folks that don't quite understand what it does, how it works or why they do/don't need it.

Given the use of it is so discouraged and that it causes issues, I'd be for removing it and then having a guide page or similar with some pointers on how to do it yourself if you really need it.

**Comment by mzrtamp** [comment]
I think the boolean for `fetchAllMembers` is already `false` as long as we don't configure it into the **ClientOptions**, and that choice depends on the user who created it ~or copy-paste from someone else's code and doesn't know what is used for~.

Sometimes this option is quite useful for those who make the bot only for 1 server, or as a dedicated servant for a server.

**Reply by kyranet** [reply]
One of the problems with that is that the built-in `fetchAllMembers` was intended to be used as a shortcut to request all members from all guilds in one WS query. However, Discord removed this feature and forced us to send a packet for each guild, which is extremely inefficient and slow (specially with the 120 packets/minute/shard ratelimit), requiring code able to handle the WS packet ratelimit to be extremely complex, so we had two options:

- Remove a troublesome option, which doesn't work for the vast majority of users (specially those in +100 guilds).
- Continue getting more and more issues regarding them, specially due to us not handling their ratelimits.

Nevertheless, it's also important to note that the `fetchAllMembers` can cause minute-long delays in hitting `ready`, if you're in large servers, this is a not-so-known thing from Discord, and generates a huge downtime in your bot.

If your bot is in a few guilds, you can listen to the `ready` event and fetch their members. If you're in very few guilds (less than 60), you can fetch them all at once without issues (unless you send a lot of voice commands).

**Reply by iCrawl** [reply]
Apart from this very extensive description of how it works:

> Sometimes this option is quite useful for those who make the bot only for 1 server, or as a dedicated servant for a server.

This is absolutely a non-issue if you just run `this.client.guilds.get(x).members.fetch()` whenever you need all of them. Which shouldn't even be a lot of use cases. This will also guarantee that you will really have **all** of them when you actually need them.

This also makes it possible to only have all members on a server you really need them. There might be certain commands disabled on a server for example. And if those disabled commands are the ones that that needs **all** members, you essentially (with `fetchAllMembers`) would have them fetched for no reason.
